pandas/core/groupby/grouper.py

```diff
@@ -563,8 +563,12 @@ def indices(self):
         if isinstance(self.grouper, ops.BaseGrouper):
             return self.grouper.indices
 
-        values = Categorical(self.grouper)
-        return values._reverse_indexer()
+        # GH 36889
+        indices_dict = {}
+        codes, uniques = algorithms.factorize(self.grouper, sort=self.sort)
+        for i, category in enumerate(Index(uniques)):
```

Comment:
can you do a dict-comprehension
remove the issue comment, but instead add a comment on what we are doing here